### PR TITLE
Fix Gan

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
## Description ##

* Remove duplicate broken GAN code from example.
* We have a working and tested GAN tutorial with MXNet in tutorials.
* Made some minor language corrections and Python compatibility.
* Added ref in the tutorials index to show GAN tutorials in mxnet site

### Essentials ###
- [X] Passed code style checking (`make lint`)
- [X] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: docs/tutorials/unsupervised_learning/gan.md [hunk 1]
```diff
@@ -1,43 +1,43 @@
-# Generative Adversarial Networks
+# Generative Adversarial Networks (GAN)
 
 GANs are an application of unsupervised learning - you don't need labels for your dataset in order to train a GAN.
- 
-The GAN framework composes of two neural networks: a generator network and a discriminator network.
 
-The generator's job is to take a set of random numbers and produce data (such as images or text).
+The GAN framework composes of two neural networks: a Generator network and a Discriminator network.
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: "composes" => "is composed of"

File: docs/tutorials/unsupervised_learning/gan.md [hunk 2]
```diff
@@ -1,43 +1,43 @@
-# Generative Adversarial Networks
+# Generative Adversarial Networks (GAN)
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: nitpick: "Generative Adversarial Network" (no s at the end) or use "GAN**s**"

File: docs/tutorials/unsupervised_learning/gan.md [hunk 3]
```diff
@@ -1,43 +1,43 @@
...
 
 GANs are an application of unsupervised learning - you don't need labels for your dataset in order to train a GAN.
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: This sentence seems slightly off.
Wikipedia's version (modified) sounds better I think:
"Generative adversarial networks (GANs) are a class of algorithms used in unsupervised learning..".

File: docs/tutorials/unsupervised_learning/gan.md [hunk 4]
```diff
@@ -1,43 +1,43 @@
...
+- How to create a custom training process in MXNet, using lower level functions from the MXNet Module API such as .bind() .forward() and .backward(). The training process for a DCGAN is more complex than many other neural networks, so we need to use these functions instead of using the higher level .fit() function.
 
 - How to visualize images as they are going through the training process
 
 ## Prerequisites
 
-This tutorial assumes you're familiar with the concept of CNN's and have implemented one in MXNet. You should also be familiar with the concept of logistic regression. Having a basic understanding for MXNet data iterators helps, since we'll create a custom Data Iterator to iterate though random numbers as inputs to our generator network. 
+This tutorial assumes you are familiar with the concepts of CNN's and have implemented one in MXNet. You should also be familiar with the concept of logistic regression. Having a basic understanding for MXNet data iterators helps, since we'll create a custom Data Iterator to iterate though random numbers as inputs to the Generator network.
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: CNN's => CNNs
"understanding for MXNet data iterators" => "understanding of MXNet data iterators"
"Data Iterator" => "data iterator"

File: docs/tutorials/unsupervised_learning/gan.md [hunk 5]
```diff
@@ -47,31 +47,31 @@ To complete this tutorial, you need:
...
+Let us start by preparing the handwritten digits from the MNIST dataset. We import the fetch_mldata function from scikit-learn, and use it to get the MNSIT dataset. Notice that it's shape is 70000x784. This contains the 70000 images on every row and 784 pixels of each image in the columns of each row. Each image is 28x28 pixels, but has been flattened so that all 784 images are represented in a single list.
 ```python
 from sklearn.datasets import fetch_mldata
 mnist = fetch_mldata('MNIST original')
 ```
 
-Next, we'll randomize the handwritten digits by using numpy to create random permutations on the dataset on our rows (images). We'll then reshape our dataset from 70000x786 to 70000x28x28, so that every image in our dataset is arranged into a 28x28 grid, where each cell in the grid represents 1 pixel of the image.
+Next, we'll randomize the handwritten digits by using numpy to create random permutations on the dataset on the rows (images). We will then reshape the dataset from 70000x786 to 70000x28x28, so that every image in the dataset is arranged into a 28x28 grid, where each cell in the grid represents 1 pixel of the image.
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: Just curious: Is there a guideline around when to use _we'll_ and when to use _we will_. In this sentence, I see both.

File: docs/tutorials/unsupervised_learning/gan.md [hunk 6]
```diff
@@ -86,25 +86,25 @@ Since the DCGAN that we're creating takes in a 64x64 image as the input, we'll u
 import cv2
 X = np.asarray([cv2.resize(x, (64,64)) for x in X])
 ```
-Each pixel in our 64x64 image is represented by a number between 0-255, that represents the intensity of the pixel. However, we want to input numbers between -1 and 1 into our DCGAN, as suggested by the research paper. To rescale our pixels to be in the range of -1 to 1, we'll divide each pixel by (255/2). This put our images on a scale of 0-2. We can then subtract by 1, to get them in the range of -1 to 1.
+Each pixel in the 64x64 image is represented by a number between 0-255, that represents the intensity of the pixel. However, we want to input numbers between -1 and 1 into the DCGAN, as suggested by the research paper. To rescale the pixels to be in the range of -1 to 1, we'll divide each pixel by (255/2). This put the images on a scale of 0-2. We can then subtract by 1, to get them in the range of -1 to 1.
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: Have we linked the research paper anywhere? If not, this would be a good place to do that.

This seems a little verbose. See if this sounds better:
"as suggested by the research paper. To rescale the pixels to be in the range of -1 to 1, we'll divide each pixel by (255/2). This put the images on a scale of 0-2. We can then subtract by 1, to get them in the range of -1 to 1." => 
"as suggested by the research paper. To rescale the pixel values, we'll divide it by (255/2). This changes the scale to 0-2. We then subtract by 1 to get them in the range of -1 to 1."

File: docs/tutorials/unsupervised_learning/gan.md [hunk 7]
```diff
@@ -86,25 +86,25 @@ Since the DCGAN that we're creating takes in a 64x64 image as the input, we'll u
...
 ```python
 X = X.astype(np.float32)/(255.0/2) - 1.0
 ```
-Ultimately, images are inputted into our neural net from a 70000x3x64x64 array, and they are currently in a 70000x64x64 array. We need to add 3 channels to our images. Typically when we are working with images, the 3 channels represent the red, green, and blue components of each image. Since the MNIST dataset is grayscale, we only need 1 channel to represent our dataset. We will pad the other channels with 0's:
+Ultimately, images are inputted into the neural net from a 70000x3x64x64 array, and they are currently in a 70000x64x64 array. We need to add 3 channels to the images. Typically when we are working with images, the 3 channels represent the red, green, and blue components of each image. Since the MNIST dataset is grayscale, we only need 1 channel to represent the dataset. We will pad the other channels with 0's:
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: I had to convince myself by reading [this](https://www.merriam-webster.com/words-at-play/is-inputted-a-word) that _inputted_ is actually a word :). 

May be paraphrasing it this way sounds better: "Ultimately, images are _**fed**_ into the neural net _**through**_ a 70000x3x64x64 array _**but**_ they are currently in a 70000x64x64 array. We need to add 3 channels to the images. ".

File: docs/tutorials/unsupervised_learning/gan.md [hunk 8]
```diff
@@ -86,25 +86,25 @@ Since the DCGAN that we're creating takes in a 64x64 image as the input, we'll u
...
 
 ```python
 X = X.reshape((70000, 1, 64, 64))
 X = np.tile(X, (1, 3, 1, 1))
 ```
-Finally, we'll put our images into MXNet's NDArrayIter, which will allow MXNet to easily iterate through our images during training. We'll also split up them images into a batches, with 64 images in each batch. Every time we iterate, we'll get a 4 dimensional array with size (64, 3, 64, 64), representing a batch of 64 images.
+Finally, we will put the images into MXNet's NDArrayIter, which will allow MXNet to easily iterate through the images during training. We will also split up them images into a batches, with 64 images in each batch. Every time we iterate, we'll get a 4 dimensional array with size (64, 3, 64, 64), representing a batch of 64 images.
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: "split up them images into a batches, with 64 images in each batch" => "split them up into batches of 64 images each"

File: docs/tutorials/unsupervised_learning/gan.md [hunk 9]
```diff
@@ -86,25 +86,25 @@ Since the DCGAN that we're creating takes in a 64x64 image as the input, we'll u
...
 ```python
 import mxnet as mx
 batch_size = 64
 image_iter = mx.io.NDArrayIter(X, batch_size=batch_size)
 ```
 ### 2. Preparing Random Numbers
 
-We need to input random numbers from a normal distribution to our generator network, so we'll create an MXNet DataIter that produces random numbers for each training batch. The DataIter is the base class of MXNet's Data Loading API. Below, we create a class called RandIter which is a subclass of DataIter. We use MXNet's built in mx.random.normal function in order to return the normally distributed random numbers every time we iterate.
+We need to input random numbers from a normal distribution to the Generator network, so we'll create an MXNet DataIter that produces random numbers for each training batch. The DataIter is the base class of MXNet's Data Loading API. Below, we create a class called RandIter which is a subclass of DataIter. We use MXNet's built in mx.random.normal function in order to return the normally distributed random numbers every time we iterate.
```
Comment (pracheer) [inline docs/tutorials/unsupervised_learning/gan.md]: Wasn't there any guideline on attaching decorations when references to code are made. E.g. DataIter, RandIter

"built in" => "built-in"

Also consider rewording:
"return the normally distributed random numbers every time we iterate." => "return numbers from a normal distribution during the iteration."

File: docs/tutorials/unsupervised_learning/gan.md [hunk 10]
```diff
@@ -117,22 +117,22 @@ class RandIter(mx.io.DataIter):
         return True
 
     def getdata(self):
-        #Returns random numbers from a gaussian (normal) distribution 
+        #Returns random numbers from a gaussian (normal) distribution
         #with mean=0 and standard deviation = 1
         return [mx.random.normal(0, 1.0, shape=(self.batch_size, self.ndim, 1, 1))]
 ```
-When we initalize our RandIter, we need to provide two numbers: the batch size and how many random numbers we want to produce a single image from. This number is referred to as Z, and we'll set this to 100. This value comes from the research paper on the topic. Every time we iterate and get a batch of random numbers, we will get a 4 dimensional array with shape: (batch_size, Z, 1, 1), which in our example is (64, 100, 1, 1).
+When we initialize the RandIter, we need to provide two numbers: the batch size and how many random numbers we want to produce a single image from. This number is referred to as Z, and we'll set this to 100. This value comes from the research paper on the topic. Every time we iterate and get a batch of random numbers, we will get a 4 dimensional array with shape: (batch_size, Z, 1, 1), which in the example is (64, 100, 1, 1).
```
Comment (madjam) [inline docs/tutorials/unsupervised_learning/gan.md]: "we want to produce a single image from" -> "we want in order to produce a single image"

File: docs/tutorials/unsupervised_learning/gan.md [hunk 11]
```diff
@@ -160,16 +160,16 @@ g5 = mx.sym.Deconvolution(gact4, name='g5', kernel=(4,4), stride=(2,2), pad=(1,1
...
+We repeat this process 4 times, applying a (2,2) stride and (1,1) pad at each Deconvolutional layer, which doubles the size of the image at each layer. By creating these layers, the Generator network will have to learn to upsample the input vector of random numbers, Z at each layer, so that network output a final image. We also reduce half the number of filters at each layer, reducing dimensionality at each layer. Ultimately, the output layer is a 64x64x3 layer, representing the size and channels of the image. We use tanh activation instead of relu on the last layer, as recommended by the research on DCGANs. The output of neurons in the final gout layer represent the pixels of generated image.
```
Comment (madjam) [inline docs/tutorials/unsupervised_learning/gan.md]: Deconvolutional -> Deconvolution

Comment (madjam) [reply docs/tutorials/unsupervised_learning/gan.md]: reduce half the number -> reduce by half the number

**Comment by sandeep-krishnamurthy** [comment]
Thank you @pracheer and @madjam for the detailed review and feedback. Made the suggested corrections.

**Comment by pracheer** [comment]
Looks good to me!
